### ISSUE TITLE
Matrix multiplication on thread broken in 1.6.1

### ISSUE THREAD

**Issue by pd0wm** [sympy/sympy#19809]
When performing matrix multiplication on a thread in version 1.6.1 I get the following error.
```
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/home/willem/.pyenv/versions/3.8.4/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/home/willem/.pyenv/versions/3.8.4/lib/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "./test.py", line 9, in main
    m = m * m
  File "/home/willem/.local/share/virtualenvs/openpilot-0b_8JeUU/lib/python3.8/site-packages/sympy/core/decorators.py", line 127, in binary_op_wrapper
    return func(self, other)
  File "/home/willem/.local/share/virtualenvs/openpilot-0b_8JeUU/lib/python3.8/site-packages/sympy/matrices/common.py", line 2494, in __mul__
    return self.multiply(other)
  File "/home/willem/.local/share/virtualenvs/openpilot-0b_8JeUU/lib/python3.8/site-packages/sympy/matrices/common.py", line 2508, in multiply
    isimpbool = _get_intermediate_simp_bool(False, dotprodsimp)
  File "/home/willem/.local/share/virtualenvs/openpilot-0b_8JeUU/lib/python3.8/site-packages/sympy/matrices/utilities.py", line 45, in _get_intermediate_simp_bool
    return _get_intermediate_simp(default, False, True, dotprodsimp)
  File "/home/willem/.local/share/virtualenvs/openpilot-0b_8JeUU/lib/python3.8/site-packages/sympy/matrices/utilities.py", line 34, in _get_intermediate_simp
    if dotprodsimp is False or _dotprodsimp_state.state is False:
AttributeError: '_thread._local' object has no attribute 'state'
```

Minimal broken example:
```python
from threading import Thread
import sympy as sp

def main():
  m = sp.Matrix([[1]])
  m = m * m

if __name__ == "__main__":
  r = Thread(target=main)
  r.start()
  r.join()
```

The error looks similar #19786, but for me version 1.6 works fine. Is there more information I can provide?

Environment:
 - System: Ubuntu 19.10
 - Python: 3.8.4 installed through pyenv
 - Sympy: Version 1.6.1 installed through pip

**Comment by oscarbenjamin** [comment]
Thanks. I'm thinking about releasing 1.6.2 and a fix for this could go in. It seems that the dotprodsimp flag is using a thread-local so it should be able to work with threading but I guess there's something wrong with it. @Pristine-Cat

**Comment by pd0wm** [comment]
Looking through the history, https://github.com/sympy/sympy/commit/ffa63442032226850a9ba3d532d68a1d74e10eaf#diff-53519477a60497f2b48b2636fc889e71 looks suspicous.

**Comment by oscarbenjamin** [comment]
How are you supposed to set the value of a thread-local for newly created threads?

**Comment by pd0wm** [comment]
I think in this case you're supposed to subclass threading.local, and make sure it sets the desired default values in the `__init__` function.

**Comment by oscarbenjamin** [comment]
Yes, I think that looks good. I don't have much experience with thread-locals...

Do these others look okay to you?
https://github.com/sympy/sympy/blob/e20ad30490ff1eb73b577a1fc07ca46323067c0a/sympy/core/parameters.py#L7

**Comment by oscarbenjamin** [comment]
Can you also open a PR against the 1.6 branch so that it can go into the 1.6.2 release?

**Comment by pd0wm** [comment]
> Yes, I think that looks good. I don't have much experience with thread-locals...
> 
> Do these others look okay to you?
> https://github.com/sympy/sympy/blob/e20ad30490ff1eb73b577a1fc07ca46323067c0a/sympy/core/parameters.py#L7

That looks similar. If you want, I can also reuse that same class instead of creating a new one.

**Comment by oscarbenjamin** [comment]
I think it's fine to have a separate class as long as they both work. We could move the dotprodsimp parameter to be with the other two but I'm not sure we want to keep it longer term.

**Comment by oscarbenjamin** [comment]
Since the fix for this is now merged both to master and the 1.6 branch I'll close this issue. Thanks for reporting and for fixing this!
